File: agent-driven-architecture-main/examples/typescript/src/spine/pure/staged.ts
// ── spine/pure/staged — the OFF-BUS INPUT fixture, sealed (5.4, 10.2, 11.2) ─
// Everything a step consumed that did NOT come off the bus, in the order it was
// staged. Two kinds, and they differ in exactly one way that matters: where the
// text came from. NEITHER IS TRUSTED.
//
//   Perceived   10.2 — content a sensor, an inbox or a webhook perceived.
//   Recalled    11.2 — a PEER TIER's published conclusion, fetched through the
//                      relay. A suggestion, never a command. It carries no
//                      Authority, and there is no field on it that could carry
//                      one, so recall confers no authority BY CONSTRUCTION.
//
// G12, expressed natively: a shared base interface declares `source` ONCE, the
// union is closed here, and every consumer matches it with a `never` guard.
//
// WHY THE RECALL OUTCOME IS A SEALED SET AND NOT A NULLABLE STRING. "The deep
// tier has not published yet" and "the deep tier is slow and this is what we
// held last" are different facts with different consequences, and a caller that
// cannot tell them apart will eventually present stale as fresh. Three variants,
// three `never`-guarded consumers (`renderStaged` below, plus the analysis
// block's view row and context line), so a fourth cannot slip past.
//
// WHY `publishedAt` AND NOT A PRE-COMPUTED `age`. An age captured at read time
// is a SECOND clock reading, unrecorded, and it would diverge on every re-fold.
// `publishedAt` is a value that travels; the age is DERIVED at the consuming
// step (`ageOf`) from the one clock read the boundary already makes. That keeps
// "`clock.now()` is the only clock read in the system" literally true, and makes
// the age deterministic under replay.

import type { SourceKey, SourceName, Timestamp } from "./ids";

// ── The relay's own record ──────────────────────────────────────────────────
// What an append-only relay hands back. Declared here, in `spine/pure`, because
// it rides `Recall` onto the committed record; the port that reads it lives in
// `spine/ports/relay`.

export interface RelayEntry {
  readonly publishedAt: Timestamp;
  readonly text: string;
}

/** How long the FAST tier will wait for the relay before degrading. The party
 *  that must not block owns the bound — never the port (11.2). */
export const RECALL_DEADLINE_MS = 50;

// ── Recall: the sealed outcome of one bounded read ──────────────────────────

export type RecallKind = "Fresh" | "LastKnown" | "Empty";

/** The parent declares both shared properties exactly once, so no variant can
 *  be added that forgets to say what it holds or when it was published. */
export interface RecallBase {
  readonly kind: RecallKind;
  readonly text: string;
  readonly publishedAt: Timestamp | null;
}

/** The read completed inside its deadline. */
export interface Fresh extends RecallBase {
  readonly kind: "Fresh";
  readonly publishedAt: Timestamp;
}

/** The deadline blew; this is the newest entry we already held. STALE, and it
 *  says so in its own type — a caller cannot mistake it for `Fresh` without
 *  writing the mistake down. */
export interface LastKnown extends RecallBase {
  readonly kind: "LastKnown";
  readonly publishedAt: Timestamp;
}

/** Wired, nothing to give. Also covers "timed out and we have never read
 *  successfully" — decided, not overlooked: the fast tier's behaviour is
 *  identical in both cases (it has no conclusion), and the operational signal
 *  that the relay is slow is the FOLDED fault, not this type. */
export interface EmptyRecall extends RecallBase {
  readonly kind: "Empty";
  readonly text: "";
  readonly publishedAt: null;
}

export type Recall = Fresh | LastKnown | EmptyRecall;

export const emptyRecall: EmptyRecall = { kind: "Empty", text: "", publishedAt: null };

export function fresh(entry: RelayEntry): Fresh {
  return { kind: "Fresh", text: entry.text, publishedAt: entry.publishedAt };
}

export function lastKnown(entry: RelayEntry): LastKnown {
  return { kind: "LastKnown", text: entry.text, publishedAt: entry.publishedAt };
}

/** Derived at the consuming step from the ONE injected clock read — never
 *  captured at read time. `null` when there is nothing to age. */
export function ageOf(recall: Recall, now: Timestamp): number | null {
  return recall.publishedAt === null ? null : now - recall.publishedAt;
}

// ── StagedInput: the sealed set the committed record carries ────────────────

export type StagedKind = "Perceived" | "Recalled";

export interface StagedInputBase {
  readonly kind: StagedKind;
  readonly source: SourceName;
}

/** Untrusted perceived content staged for this turn (10.2).
 *
 *  `key` is the source's OWN id for this work item, and it rides the COMMITTED
 *  record on purpose: the durable policy's dedupe scope is rebuilt from the
 *  timeline at recovery (12.2), so "each work item folds at most once" survives
 *  a process restart. A key held only on the uncommitted mailbox envelope dies
 *  with the process — which is exactly the double-fold this field closes. */
export interface Perceived extends StagedInputBase {
  readonly kind: "Perceived";
  readonly body: string;
  readonly key: SourceKey;
}

/** A peer tier's conclusion, staged for this turn (11.2). There is exactly ONE
 *  production site for this variant in the system — the consumer — mirroring
 *  the one-production-site rule `ToolResult` already lives under (C7). A
 *  producer cannot post a forged relay snapshot. */
export interface Recalled extends StagedInputBase {
  readonly kind: "Recalled";
  readonly recall: Recall;
}

export type StagedInput = Perceived | Recalled;

/** The discriminant as a CONSTANT, and the ONE place it is compared — the same
 *  move `spine/pure/mailbox` makes for `Input`. Recovery filters a committed
 *  timeline through this guard to rebuild the durable dedupe scope. */
export const PERCEIVED_KIND = "Perceived" as const;

export function isPerceived(staged: StagedInput): staged is Perceived {
  return staged.kind === PERCEIVED_KIND;
}

export function perceived(source: SourceName, body: string, key: SourceKey): Perceived {
  return { kind: "Perceived", source, body, key };
}

export function recalled(source: SourceName, recall: Recall): Recalled {
  return { kind: "Recalled", source, recall };
}

/** CONSUMER 1 of 3 of the sealed `Recall`. The digest the model reads must say
 *  which branch was taken, because that digest is what the committed
 *  `ContextFixture` pins — a live run that recalled `Fresh` and a re-fold that
 *  resolved `LastKnown` would produce different text and fail the golden trace. */
export function renderStaged(staged: StagedInput): string {
  switch (staged.kind) {
    case "Perceived":
      return `staged[${staged.source}]: ${staged.body}`;
    case "Recalled":
      return `recall[${staged.source}]: ${renderRecall(staged.recall)}`;
    default: {
      const _never: never = staged;
      return _never;
    }
  }
}

function renderRecall(recall: Recall): string {
  switch (recall.kind) {
    case "Fresh":
      return `fresh, published @${recall.publishedAt} — ${recall.text}`;
    case "LastKnown":
      return `LAST KNOWN (relay did not answer in time), published @${recall.publishedAt} — ${recall.text}`;
    case "Empty":
      return "no conclusion published";
    default: {
      const _never: never = recall;
      return _never;
    }
  }
}

// ── THE TURN'S SCOPE — what a tool's `execute` is handed by the CALL ─────────
// `staged` varies per turn; the tool table does not. While the table closed over
// `staged`, the table had to be rebuilt every turn — which is exactly what kept
// this port from declaring ONE agent and reusing it (SDK-3), the shape
// `ToolLoopAgent` exists for. The SDK's answer is to pass per-call values to
// `execute` through the call's context, so the table stops varying.
//
// THE NARROWING LIVES HERE AND NOT IN THE LOOP, and the gate is what said so:
// C14 refused a ternary in `spine/agent/loop`, on the grounds that "a decision
// the loop may not make — decisions belong to the fold". It was right. The seam
// hands `unknown`, so SOMETHING has to decide what an absent scope means; that
// decision is about staged input, so it belongs to the module that owns staged.

export interface TurnScope {
  readonly staged: readonly StagedInput[];
}

const EMPTY_SCOPE: TurnScope = { staged: [] };

/** A call that supplied no scope reads as the EMPTY turn, never as a crash: the
 *  spine does not throw at a seam (spine/boundary/action states the same rule). */
export function scopeOf(context: unknown): TurnScope {
  const staged = (context as { staged?: readonly StagedInput[] } | null)?.staged;
  if (staged === undefined) return EMPTY_SCOPE;
  return { staged };
}

File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/pure/Staged.kt
// ── spine/pure/staged — every OFF-BUS input, as one closed set (5.4, 11.2) ─
// A step consumes things that did not come off the bus. There are exactly two
// kinds, and G12 says a closed set is sealed with its shared property declared ONCE
// on the parent:
//
//     Perceived   10.2 — untrusted content the world offered (a ticket, a webhook)
//     Recalled    11.2 — a peer tier's published conclusion, reached through recall
//
// Both carry `source`, so the committed record always says WHERE a staged value
// came from, and neither can be confused for the other at a consumer.
//
// Why this is one list and not two capture slots: 4.7 forbids a block a private
// capture log, and 5.4 already specifies plural off-bus inputs "in their staging
// order, keyed to the consuming step". StepRecord.staged IS that slot — the bus
// supplies the order and the step index is the key. ORDER IS PINNED:
// [Perceived?, Recalled?], perception first. The order reaches the rendered digest
// and therefore the committed ContextFixture, so it is law, not style.
//
// ZERO I/O. No async, no clock, no coroutines — gate check C8.

package adr.spine.pure

/** How long the fast tier will wait for the deep tier before degrading (11.2). */
const val RECALL_DEADLINE_MS: Millis = 50

/** One off-bus input, staged for exactly one step and captured on its record. */
sealed class StagedInput(
    /** WHERE it came from. Declared once; carried by every variant by construction. */
    open val source: SourceName,
) {

    /**
     * UNTRUSTED perceived content (10.2). Data to reason about, never an instruction.
     *
     * [key] is the source's OWN id for this work item, and it rides the COMMITTED
     * record on purpose: the durable policy's dedupe scope is rebuilt from the
     * timeline at recovery (12.2), so "each work item folds at most once" survives a
     * process restart. A key held only on the uncommitted mailbox envelope dies with
     * the process — which is exactly the double-fold this field closes.
     */
    data class Perceived(
        override val source: SourceName,
        val body: String,
        val key: SourceKey,
    ) : StagedInput(source)

    /**
     * A peer tier's conclusion, reached through the recall verb (11.2/11.3).
     *
     * Equally untrusted: recall confers NO authority. A recalled conclusion is a
     * suggestion, and the boundary gate still judges any irreversible act on its
     * own merits — see §5 of the design note and RelayTest's two injection cases.
     */
    data class Recalled(override val source: SourceName, val recall: Recall) : StagedInput(source)
}

/**
 * What a bounded recall actually returned. Three variants, and the whole point is
 * that STALE IS NEVER PRESENTED AS FRESH: the degrade is a different type, not a
 * flag on the same one, so every consumer is forced by the compiler to say what it
 * does about it.
 *
 * `publishedAt` and not a pre-computed `age`: an age captured at read time is a
 * SECOND, unrecorded clock reading, and it would diverge on every re-fold.
 * `publishedAt` is a value that travels; the age is derived at the consuming step
 * from the one clock read the boundary already makes (G9), so it replays exactly.
 */
sealed class Recall(
    /** Every outcome has text — `Empty`'s is empty. Declared once, HELD once (G12). */
    open val text: String,
    /** The entry's own clock reading, or null when there is no entry. Declared once. */
    open val publishedAt: Timestamp?,
) {

    /** The read completed inside its deadline. */
    data class Fresh(override val text: String, override val publishedAt: Timestamp) :
        Recall(text, publishedAt)

    /** The deadline blew; this is the newest entry the reader already held. */
    data class LastKnown(override val text: String, override val publishedAt: Timestamp) :
        Recall(text, publishedAt)

    /**
     * Wired, nothing to give: the deep tier has not published, or it timed out and
     * nothing was ever read successfully. One variant on purpose — the fast tier's
     * behaviour is identical in both cases (it has no conclusion), and the signal
     * that a relay is slow is a folded fault, not a fourth variant nobody branches on.
     */
    data object Empty : Recall(text = "", publishedAt = null)
}

/** One published conclusion on the append-only relay. Plain data; no handle, no method. */
data class RelayEntry(val publishedAt: Timestamp, val text: String)
